### ISSUE TITLE
add support in Pool retry to run again  the query if the database is restarted

### ISSUE THREAD

**Issue by insinfo** [isoos/postgresql-dart#290]
It would be very useful to have a retry if the query throws an error if the database is restarted

**Comment by isoos** [comment]
Yes, I'm planning to do this, however, I'm not sure what the best API choice would be:
- putting the extra parameter(s) in `Pool.run[Tx]()` https://pub.dev/documentation/postgres/latest/postgres/Pool/run.html
- putting the extra parameter(s) in `Session.execute` (and prepare+run)
- putting the extra parameter(s) into one of the `*Settings` object

Any insight on how this feels better? /cc @simolus3

**Comment by simolus3** [comment]
> Any insight on how this feels better? /cc @simolus3

Given that this is likely pool-specific, putting it on `Session.execute` feels a bit weird. I also think this should be part of `PoolSettings` (perhaps with an optional parameter on `Pool.run` to override the default behavior in a limited scope, but I'm not sure how useful that would be).

**Comment by isoos** [comment]
Iterating a bit more on this: we may split the connection-related retries (which in the pool context may just land on a different instance) and the application-logic retries (which may be run on the same instance).

1) I think the connection-related retry settings may go to into the `PoolSettings` object, as it will be applied over the entire lifecycle of the pool usage. I'm unsure if we may allow an override it in a potential `PoolSessionSettings` or `PoolTransactionSettings` object, but that is certainly a future possibility there.

2) I think the application-logic retries should not go to the `PoolSettings` object, or even if they go, we must have a `run[Tx]`-level override for them.

3) If these are implemented separately, we may retry an operation `N * M` times (assuming `N` connection retry and `M` application-level retry). This may not be bad, just a side-effect that we need to document and be explicit about. Maybe it is easier to merge the two options and have only `max(N, M)` retries though, over a single loop. I'm going back and forth on what is better, feel free to weight any opinion on that.

4) I'm also a bit on the side to not expose `package:retry` APIs directly on the `package:postgres` API, only to rely on it in the implementation. Exposing fewer options helps to keep the complexity somewhat lower.

**Comment by busslina** [comment]
Hi, is there any example of a Pool usage available?

**Comment by isoos** [comment]
> Hi, is there any example of a Pool usage available?

What are you looking for? It is mostly a drop-in for connections and their renewals, so there is not much to it in terms of how to use them.

Maybe the tests can give you further hints?
https://github.com/isoos/postgresql-dart/blob/master/test/pool_test.dart

**Comment by busslina** [comment]
I reformulated the question [here](https://github.com/isoos/postgresql-dart/discussions/301)

**Comment by ryanheise** [comment]
Just to mention, there is an "unreferenced" `Retry` class in the API (`lib/postgres.dart`):

```
/// The settings that control the retry of [SessionExecutor.run] and [SessionExecutor.runTx] methods.
class Retry<R> {
  /// The maximum number of attempts to run the operation.
  final int maxAttempts;

  final FutureOr<R> Function()? orElse;
  final FutureOr<bool> Function(Exception)? retryIf;

  Retry({
    required this.maxAttempts,
    this.orElse,
    this.retryIf,
  });
}
```

(I don't have any ideas to contribute on what the API could be, except that implementing exponential backoff would be nice.)

**Comment by isoos** [comment]
@ryanheise Yes, that's something I wanted to include in the `run` / `runTx` methods, although I think it was stalled because of an API indecision (whether it becomes an extra parameter or part of the `*Settings` object, and if the later, how does it make sense in all context (incl. pool connection retry vs. run fn retry). Any preferences or opinions there?

**Comment by isoos** [comment]
> wouldn't it make sense to have this logic in withConnection()?

In case of the pool, `run` and `runTx` are calls to `withConnection()`, so there is not much difference there, and in the non-pool case they could be still meaningful in case of transaction or optimistic query retries.

**Comment by hurrba** [comment]
for anyone intereseted, I ended up creating an extension as I also wanted to handle other exceptions:

```
extension SessionExtension on Session {
  Future<Result> executeAndCatch(
    String sql, {
    Object? parameters,
  }) async {
    return await _execute(sql, parameters: parameters);
  }

  Future<Result> _execute(
    String sql, {
    Object? parameters,
    int retryCount = 0,
  }) async {
    try {
      return await execute(Sql.named(sql), parameters: parameters);
    } catch (e) {
      switch (e) {
        case ServerException(code: '23505', message: final message):
          throw DuplicateKeyException(message);
        case ServerException(code: '23503', message: final message):
          throw ForeignKeyViolationException(message);
        case PgException(message: 'Attempting to execute query, but connection is not open.', severity: Severity.error) ||
             SocketException(osError: OSError(message: 'Connection refused', errorCode: 111)):
          if (retryCount < 3) {
            int backoff = 1 << retryCount;

            logger.info('Connection is not open, retrying in $backoff seconds...');

            await Future.delayed(Duration(seconds: backoff));
            return await _execute(
              sql,
              parameters: parameters,
              retryCount: retryCount + 1,
            );
          }
          rethrow;
        default:
          logger.severe('Error executing query:\n$sql', e);
          rethrow;
      }
    }
  }
}
```

**Comment by isoos** [comment]
I'm curious: we have had a lot of bug fixes in the recent months, where are we standing with this issue? Is there any suggestion we could adopt?

@hurrba: I like the `ForeignKeyViolationException` and `DuplicateKeyException` subtypes, would you be interested to submit them as PRs?

**Comment by hurrba** [comment]
@isoos sure. are you thinking something along the lines of how you handle `_PgQueryCancelledException`  [here](https://github.com/isoos/postgresql-dart/blob/75637139f35c2ea01e045a5f1301c8970f13dd6d/lib/src/exceptions.dart#L193)?

**Comment by isoos** [comment]
@hurrba: Yes, that would be the ideal place for it.

**Comment by hurrba** [comment]
> [@hurrba](https://github.com/hurrba): Yes, that would be the ideal place for it.

https://github.com/isoos/postgresql-dart/pull/416
